README.md
# ProcessWire newsletter subscription

Compatibility: ProcessWire 3.x

Allow users to subscribe and unsubscribe to a newsletter.

## Subscription flow

The user enters their email address and some other information (if you want to).
As soon as the form is submitted a mail is sent to the given address containing a confirmation link (double opt-in).
Now the user is added with status hidden and role newsletter.
After visiting the confirmation link the user is subscribed (changed status to active).

## Unsubscription flow

There are two methods to unsubscribe: by using a web form or by providing a link in the newsletter.

### A) using the form

The user enters only their email address and selects the unsubscribe option (if you provide additional fields hide them via JavaScript).
If the user submitted a valid and existing email address, a email is sent to this address containing a confirmation link to end subscription.
After visiting the link the user is unsubscribed (the user will be deleted).

### B) link inside newsletter

The user is unsubscribed after visiting the link immediately.
You have to generate the link parameters depending on email and userAuthSalt.

**Example:** `https://domain.com/newsletter/?a=unsubscribe&t={sha1($email . $this->config->userAuthSalt)}`

## Usage

* install the module
* ~~create fields which you want to add to the newsletter subscription form~~
  * if the field should be required, add this in the field settings
  * field email is mandatory and already preselected
* assign these fields to the user template 
  * [How to assign fields to the user template](https://processwire.com/talk/topic/1156-custom-user-fields/?p=10161)
* fill module settings
  * ~~select fields which should be added to the form - save module config~~
  * select fields which should be used as username
  * ~~if you want to provide an unsubscribe option, activate the according checkbox~~
  * enter email from address
  * enter how many days the confirmation links should be valid (default 5 days)
  * ~~define email text messages~~ Now HTML-Templates
* call module
  
  
  ```php
  echo $modules->get('NewsletterSubscription')->watch();
  ```
 
  

## Notify admin via email

If you want to notify any person e.g. administator via email when an user has subscribed/unsubscribed,
check the according checkbox in module settings.
As soon as you activate this field, some more fields will appear.
In these fields you can specify further email messages as well as email receiver and sender.
Furthermore you can use placeholders like `%email%` - even for the receiver email addresses.
After the user has visited the confirmation link the admin gets notified.

## How to translate

All phrases like email subjects are translatable. 
Add the module file to the language translator and start translating.
Phrases which don't exist in this file belong to the ProcessWire core.
For example the messages *Please enter a valid e-mail address* or *Missing required value*.

Relevant Files:

- site/modules/NewsletterSubscription.module
- wire/modules/Inputfield/InputfieldEmail.module
- wire/core/InputfieldWrapper.php

Depending on the fields you added to the form there might be some other files.


CHANGELOG.md
# Changelog

### 0.0.4 (2017-07)

- Removed forms, so the plugin watches all posted values and does his work if the reqiuered values are present
- removed unsubscribe-option because there is no auto-generated form


### 0.0.3 (2016-02-19)

- extends option array
  - adds prependMarkup: optional markup to prepend to the form output
  - adds appendMarkup: optional markup to append to the form output

### 0.0.2 (2016-02-07)

- adds hidden field `roles` to enable *requiredIf roles=newsletter* dependencies

### 0.0.1 (2016-02-03)

- initial module


NeleSubscriptionConfig.php
<?php namespace ProcessWire;

/**
 * Class NewsletterSubscriptionConfig
 */
class NeleSubscriptionConfig extends ModuleConfig {

	/**
	 * array Default config values
	 */
	public function getDefaults() {
		return array(
			'formfields' => 'email',
			'namefields' => '',
			'unsubscribe' => false,
			'noCompare' => false,
			'mailfrom' => 'noreply@server.com',
			'periodOfValidity' => 5,
			'messageSubscribe' => "Hey %name%\n\nPlease click the following link to confirm your subscription: %link%",
			'messageUnsubscribe' => "Hey %name%\n\nPlease click the following link to end your subscription: %link%",
			'notifyAdmin' => 0,
			'notifyAdminMailto' => 'mailto@server.com',
			'notifyAdminSubscribeMailfrom' => 'noreply@server.com',
			'notifyAdminUnsubscribeMailfrom' => 'noreply@server.com',
			'notifyAdminMessageSubscribe' => "New newsletter subscription",
			'notifyAdminMessageUnsubscribe' => "New newsletter unsubscription",
		);
	}

	/**
	 * Retrieves the list of config input fields
	 * Implementation of the ConfigurableModule interface
	 *
	 * @return InputfieldWrapper
	 */
	public function getInputfields() {
		$formfields = $this->data['formfields'];
		$unsubscribe = $this->data['unsubscribe'];
		$notifyAdmin = $this->data['notifyAdmin'];
		$noCompare = $this->data['noCompare'];
		$inputfields = parent::getInputfields();

		// select fields
		$field = $this->modules->get('InputfieldAsmSelect');
		$field->description = __('Add all fields which should be attached to the subscription form.') . PHP_EOL .
			__('You have to add all needed fields to the user template first¹.');
		$field->addOption('', '');
		$field->label = __('Select form fields for subscription');
		$field->attr('name', 'formfields');
		$field->required = true;
		$field->columnWidth = 50;
		foreach (wire('templates')->get('user')->fields as $f) {
			$field->addOption($f->name, $f->name);
		}
		$inputfields->add($field);

		// select name fields
		$field = $this->modules->get('InputfieldAsmSelect');
		$field->description = __('Add the fields which should be used as username.') . PHP_EOL .
			__('Please save the chosen form fields first to get a list to select from.');
		$field->label = __('Select name fields');
		$field->attr('name', 'namefields');
		$field->columnWidth = 50;
		if (is_array($formfields)) {
			foreach ($formfields as $formfield) {
				$f = wire('fields')->get($formfield);
				$field->addOption($f->name, $f->name);
			}
		}
		$inputfields->add($field);

		// help - how to add fields to the user template
		$help = $this->modules->get('InputfieldMarkup');
		$helpLink = 'https://processwire.com/talk/topic/1156-custom-user-fields/?p=10161';
		$helpContent = '<p><a  target="_blank" href="' . $helpLink . '"><sup>1</sup> How to add fields to the user template?</a></p>';
		$help->value = $helpContent;
		$inputfields->add($help);

		/*// unsubscribe field
		$field = $this->modules->get('InputfieldCheckbox');
		$field->name = 'unsubscribe';
		$field->label = 'Unsubscribe option';
		$field->description = __('Should the form contain an unsubscribe option?');
		$field->value = 1;
		$field->attr('checked', $unsubscribe ? 'checked' : '');
		$field->columnWidth = 25;
		$inputfields->add($field);*/

		// noCompare field
		$field = $this->modules->get('InputfieldCheckbox');
		$field->name = 'noCompare';
		$field->label = 'Do not compare E-Mail-Addresses';
		$field->description = __('Do not check whether an E-Mail-Address already exists.');
		$field->value = 1;
		$field->attr('checked', $noCompare ? 'checked' : '');
		$field->columnWidth = 25;
		$inputfields->add($field);

		// mailserver field
		$field = $this->modules->get('InputfieldText');
		$field->name = 'mailfrom';
		$field->label = __('E-Mail From Address');
		$field->description = __('Sender Address');
		$field->columnWidth = 25;
		$inputfields->add($field);

		// periodOfValidity field
		$field = $this->modules->get('InputfieldInteger');
		$field->name = 'periodOfValidity';
		$field->label = __('Period of Validity');
		$field->description = __('Number of days confirmation links are valid.');
		$field->columnWidth = 25;
		$inputfields->add($field);

		/*// new fieldset containing messages
		$fieldset = $this->modules->get('InputfieldFieldset');
		$fieldset->label = $this->_('E-Mail Messages');
		$fieldset->collapsed = Inputfield::collapsedYes;
		$fieldset->icon = 'send';

		$field = $this->modules->get('InputfieldTextarea');
		$field->name = 'messageSubscribe';
		$field->label = __('Subscribe E-Mail Message');
		$field->description = __('Use %fieldName% as placeholder, for example %email%');
		$field->rows = 8;
		$field->columnWidth = 50;
		$fieldset->add($field);

		$field = $this->modules->get('InputfieldTextarea');
		$field->name = 'messageUnsubscribe';
		$field->label = __('Unsubscribe E-Mail Message');
		$field->description = __('Use %fieldName% as placeholder, for example %email%');
		$field->rows = 8;
		$field->columnWidth = 50;
		$fieldset->add($field);

		$inputfields->add($fieldset);*/

		// new fieldset containing admin notification
		$fieldset = $this->modules->get('InputfieldFieldset');
		$fieldset->label = $this->_('Notify Admin');
		$fieldset->collapsed = Inputfield::collapsedYes;
		$fieldset->icon = 'comment';

		// notify admin field
		$field = $this->modules->get('InputfieldCheckbox');
		$field->name = 'notifyAdmin';
		$field->label = 'Notify Admin';
		$field->description = __('Should the admin be notified by email?');
		$field->value = 1;
		$field->columnWidth = 50;
		$field->attr('checked', $notifyAdmin ? 'checked' : '');
		$fieldset->add($field);

		// notify admin - mailto field
		$field = $this->modules->get('InputfieldText');
		$field->name = 'notifyAdminMailto';
		$field->label = __('Admin Notification E-Mail');
		$field->description = __('Mailto Address');
		$field->columnWidth = 50;
		$field->showIf = "notifyAdmin=1";
		$fieldset->add($field);

		// notify admin - subscribtion - mailserver field
		$field = $this->modules->get('InputfieldText');
		$field->name = 'notifyAdminSubscribeMailfrom';
		$field->label = __('Subscription - E-Mail From Address');
		$field->description = __('Sender Address');
		$field->columnWidth = 50;
		$field->showIf = "notifyAdmin=1";
		$fieldset->add($field);

		// notify admin - unsubscribtion - mailserver field
		$field = $this->modules->get('InputfieldText');
		$field->name = 'notifyAdminUnsubscribeMailfrom';
		$field->label = __('Unsubscription - E-Mail From Address');
		$field->description = __('Sender Address');
		$field->columnWidth = 50;
		$field->showIf = "notifyAdmin=1";
		$fieldset->add($field);

		// notify admin - subscription - message field
		$field = $this->modules->get('InputfieldTextarea');
		$field->name = 'notifyAdminMessageSubscribe';
		$field->label = __('Subscribe E-Mail Message');
		$field->description = __('Use %fieldName% as placeholder, for example %email%');
		$field->rows = 8;
		$field->columnWidth = 50;
		$field->showIf = "notifyAdmin=1";
		$fieldset->add($field);

		// notify admin - unsubscription - message field
		$field = $this->modules->get('InputfieldTextarea');
		$field->name = 'notifyAdminMessageUnsubscribe';
		$field->label = __('Unsubscribe E-Mail Message');
		$field->description = __('Use %fieldName% as placeholder, for example %email%');
		$field->rows = 8;
		$field->columnWidth = 50;
		$field->showIf = "notifyAdmin=1";
		$fieldset->add($field);

		$inputfields->add($fieldset);

		return $inputfields;
	}

}
